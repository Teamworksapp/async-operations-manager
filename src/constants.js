import {
  makeConstantsObject,
} from './helpers';

const readAsyncOperationFieldsToPullFromParent = [
  'fetchStatus',
  'dataStatus',
  'lastFetchStatusTime',
  'lastDataStatusTime',
];

const ASYNC_OPERATION_STEPS = makeConstantsObject([
  'BEGIN_ASYNC_OPERATION',
  'RESOLVE_ASYNC_OPERATION',
  'REJECT_ASYNC_OPERATION',
]);

const ASYNC_OPERATION_TYPES = makeConstantsObject([
  'READ',
  'WRITE',
]);

/**
 * In general this is used to note the status of the last fetch attempted for a backend operation.
 */
const FETCH_STATUS = makeConstantsObject([
  // Default: we haven't tried to fetch yet
  'NULL',
  // Indicates that a request is in flight.
  'PENDING',
  // Indicates that the request completed and all is well
  'SUCCESSFUL',
  // Indicates that something went wrong: the request didn't complete, or it completed with
  // an http, payload/parsing, or business logic error.
  // @TODO: Should we distinguish between "request completed by result failed" vs
  //        "request didn't complete due to timeout or network error"?
  'FAILED',
  // @TODO: Is it worth adding 'SKIPPED' or 'OFFLINE' for cases where we were asked to
  //        do a fetch but didn't even try?
]);


/**
 * In general this is used to note the presence of data for a specific backend operation.
 */
const DATA_STATUS = makeConstantsObject([
  // Default: we haven't received anything
  'ABSENT',
  // Result was received and we have data
  'PRESENT',
  // @TODO: Is it worth tracking an OFFLINE-like state?
]);

const WILDCARD = '*';

export {
  readAsyncOperationFieldsToPullFromParent,

  ASYNC_OPERATION_TYPES,
  ASYNC_OPERATION_STEPS,
  FETCH_STATUS,
  DATA_STATUS,
  WILDCARD,
};
