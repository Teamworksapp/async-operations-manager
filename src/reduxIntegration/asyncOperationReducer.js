import { includes } from 'lodash';

import {
  getStateForOperationAfterStep,
  getAsyncOperationsManagerState,
} from '../asyncOperationManagerUtils';

import {
  ASYNC_OPERATION_STEPS,
} from '../constants';

function asyncOperationReducer(state = getAsyncOperationsManagerState(), action) {
  if (includes(ASYNC_OPERATION_STEPS, action.operationStep) && action.descriptorId) {
    return {
      ...state,
      ...getStateForOperationAfterStep(state, action.operationStep, action.descriptorId, action),
    };
  }
  return state;
}


export default asyncOperationReducer;
