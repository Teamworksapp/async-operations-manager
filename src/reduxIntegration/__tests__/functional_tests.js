/* eslint-env jest */
import { expect } from 'chai';

import asyncOperationReducer from '../asyncOperationReducer';

import { FETCH_STATUS, DATA_STATUS } from '../../constants';

import {
  createAsyncOperationInitialAction,
  createAsyncOperationBeginAction,
  createAsyncOperationResolveAction,
  createAsyncOperationRejectAction,
} from '../asyncOperationReduxUtils';

import {
  clearAsyncOperationsManagerState,
  getAsyncOperationsManagerState,
  registerAsyncOperationDescriptors,
  invalidateAsyncOperation,
} from '../../asyncOperationManagerUtils';

describe('functional tests', () => {
  let state;

  beforeEach(() => {
    clearAsyncOperationsManagerState();
    const dateNowStub = jest.fn(() => 1530518207007);
    global.Date.now = dateNowStub;
  });

  describe('READ operation scenarios', () => {
    let initialAction;
    let beginAction;
    beforeEach(() => {
      state = {};
      clearAsyncOperationsManagerState();
      registerAsyncOperationDescriptors({
        descriptorId: 'FETCH_PERSON_DATA',
        requiredParams: ['personId'],
        operationType: 'READ',
      });
      initialAction = createAsyncOperationInitialAction('FETCH_PERSON_DATA', {
        personId: 111,
      });
      beginAction = createAsyncOperationBeginAction('FETCH_PERSON_DATA', {
        personId: 111,
      });
    });

    it('should update a successful READ operation as expected from start to finish', () => {
      const resolveAction = createAsyncOperationResolveAction('FETCH_PERSON_DATA', {
        personId: 111,
      });

      expect(asyncOperationReducer(state, initialAction)).to.deep.equal(state);
      expect(asyncOperationReducer(state, beginAction).operations).to.deep.equal({
        FETCH_PERSON_DATA_111: {
          descriptorId: 'FETCH_PERSON_DATA',
          fetchStatus: FETCH_STATUS.PENDING,
          dataStatus: DATA_STATUS.ABSENT,
          message: null,
          lastFetchStatusTime: 1530518207007,
          lastDataStatusTime: 0,
          personId: 111,
          key: 'FETCH_PERSON_DATA_111',
        },
      });
      expect(asyncOperationReducer(state, resolveAction).operations).to.deep.equal({
        FETCH_PERSON_DATA_111: {
          descriptorId: 'FETCH_PERSON_DATA',
          fetchStatus: FETCH_STATUS.SUCCESSFUL,
          dataStatus: DATA_STATUS.PRESENT,
          message: null,
          lastFetchStatusTime: 1530518207007,
          lastDataStatusTime: 1530518207007,
          personId: 111,
          lastFetchFailed: false,
          key: 'FETCH_PERSON_DATA_111',
        },
      });
    });

    it('should update a failed READ operation as expected from start to finish', () => {
      const rejectAction = createAsyncOperationRejectAction('FETCH_PERSON_DATA', {
        personId: 111,
      });

      expect(asyncOperationReducer(state, initialAction)).to.deep.equal(state);
      expect(asyncOperationReducer(state, beginAction).operations).to.deep.include({
        FETCH_PERSON_DATA_111: {
          descriptorId: 'FETCH_PERSON_DATA',
          fetchStatus: FETCH_STATUS.PENDING,
          dataStatus: DATA_STATUS.ABSENT,
          message: null,
          lastFetchStatusTime: 1530518207007,
          lastDataStatusTime: 0,
          personId: 111,
          key: 'FETCH_PERSON_DATA_111',
        },
      });
      expect(asyncOperationReducer(state, rejectAction).operations).to.deep.include({
        FETCH_PERSON_DATA_111: {
          descriptorId: 'FETCH_PERSON_DATA',
          fetchStatus: FETCH_STATUS.FAILED,
          dataStatus: DATA_STATUS.ABSENT,
          message: null,
          lastFetchStatusTime: 1530518207007,
          lastDataStatusTime: 0,
          personId: 111,
          lastFetchFailed: true,
          key: 'FETCH_PERSON_DATA_111',
        },
      });
    });
  });

  describe('Invalidate operations', () => {
    beforeEach(() => {
      state = {};
      clearAsyncOperationsManagerState();
    });

    it('should invalidate async operation if another operation descriptor\'s onResolve invalidates it', () => {
      registerAsyncOperationDescriptors([
        {
          descriptorId: 'UPDATE_APPOINTMENT_DATA',
          requiredParams: ['orgId', 'appointmentId'],
          operationType: 'WRITE',
          onResolve: ({ orgId }) => {
            invalidateAsyncOperation('FETCH_CALENDAR_DATA', { orgId });
          },
        },
        {
          descriptorId: 'FETCH_CALENDAR_DATA',
          requiredParams: ['orgId'],
          operationType: 'READ',
        },
      ]);

      const initialFetchCalendarDataAction = createAsyncOperationInitialAction('FETCH_CALENDAR_DATA', {
        orgId: 22,
      });
      const beginFetchCalendarDataAction = createAsyncOperationBeginAction('FETCH_CALENDAR_DATA', {
        orgId: 22,
      });
      const resolveFetchCalendarDataAction = createAsyncOperationResolveAction('FETCH_CALENDAR_DATA', {
        orgId: 22,
      });

      expect(asyncOperationReducer(state, initialFetchCalendarDataAction)).to.deep.equal(state);
      expect(asyncOperationReducer(state, beginFetchCalendarDataAction).operations).to.deep.include({
        FETCH_CALENDAR_DATA_22: {
          descriptorId: 'FETCH_CALENDAR_DATA',
          fetchStatus: FETCH_STATUS.PENDING,
          dataStatus: DATA_STATUS.ABSENT,
          message: null,
          lastFetchStatusTime: 1530518207007,
          lastDataStatusTime: 0,
          orgId: 22,
          key: 'FETCH_CALENDAR_DATA_22',
        },
      });

      expect(asyncOperationReducer(state, resolveFetchCalendarDataAction).operations).to.deep.include({
        FETCH_CALENDAR_DATA_22: {
          descriptorId: 'FETCH_CALENDAR_DATA',
          fetchStatus: 'SUCCESSFUL',
          dataStatus: 'PRESENT',
          message: null,
          lastFetchStatusTime: 1530518207007,
          lastDataStatusTime: 1530518207007,
          orgId: 22,
          lastFetchFailed: false,
          key: 'FETCH_CALENDAR_DATA_22',
        },
      });

      const dateNowStub = jest.fn(() => 1540000000000);
      global.Date.now = dateNowStub;

      const initialUpdateAppointmentDataAction = createAsyncOperationInitialAction('UPDATE_APPOINTMENT_DATA', {
        orgId: 22,
        appointmentId: 111,
      });
      const beginUpdateAppointmentDataAction = createAsyncOperationBeginAction('UPDATE_APPOINTMENT_DATA', {
        orgId: 22,
        appointmentId: 111,
      });
      const resolveUpdateAppointmentDataAction = createAsyncOperationResolveAction('UPDATE_APPOINTMENT_DATA', {
        orgId: 22,
        appointmentId: 111,
      });

      expect(asyncOperationReducer(state, initialUpdateAppointmentDataAction)).to.deep.equal(state);
      expect(asyncOperationReducer(state, beginUpdateAppointmentDataAction).operations).to.deep.include({
        UPDATE_APPOINTMENT_DATA_22_111: {
          descriptorId: 'UPDATE_APPOINTMENT_DATA',
          fetchStatus: FETCH_STATUS.PENDING,
          message: null,
          lastFetchStatusTime: 1540000000000,
          orgId: 22,
          appointmentId: 111,
          key: 'UPDATE_APPOINTMENT_DATA_22_111',
        },
      });

      expect(asyncOperationReducer(state, resolveUpdateAppointmentDataAction).operations).to.deep.include({
        UPDATE_APPOINTMENT_DATA_22_111: {
          fetchStatus: 'SUCCESSFUL',
          message: null,
          lastFetchStatusTime: 1540000000000,
          descriptorId: 'FETCH_CALENDAR_DATA',
          orgId: 22,
          appointmentId: 111,
          key: 'UPDATE_APPOINTMENT_DATA_22_111',
          dataStatus: 'ABSENT',
          lastFetchFailed: false,
          lastDataStatusTime: 0,
        },
      });

      const currentState = getAsyncOperationsManagerState();


      const fetchCalendarDataAsyncOperation = currentState.operations.FETCH_CALENDAR_DATA_22;

      expect(fetchCalendarDataAsyncOperation).to.be.an('object');
      expect(fetchCalendarDataAsyncOperation).to.deep.include({
        lastFetchStatusTime: 0,
        lastDataStatusTime: 0,
      });
      expect(fetchCalendarDataAsyncOperation).to.matchSnapshot('Invalidated read fetchCalendarDataAsyncOperation');
    });
  });

  describe('WRITE operation scenarios', () => {
    beforeEach(() => {
      state = {};
      clearAsyncOperationsManagerState();
    });
  });
});
