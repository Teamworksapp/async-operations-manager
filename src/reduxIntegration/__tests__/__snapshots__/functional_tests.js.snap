// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functional tests Invalidate operations should invalidate async operation if another operation descriptor's onResolve invalidates it: Invalidated read fetchCalendarDataAsyncOperation 1`] = `
Object {
  "dataStatus": "ABSENT",
  "descriptorId": "FETCH_CALENDAR_DATA",
  "fetchStatus": "NULL",
  "key": "FETCH_CALENDAR_DATA_22",
  "lastDataStatusTime": 0,
  "lastFetchFailed": false,
  "lastFetchStatusTime": 0,
  "message": null,
  "orgId": 22,
}
`;
